### PR TITLE
Updating FastjetJetProducer (and related producers) with fillDescriptions

### PR DESCRIPTION
**Pull request by alefisico**
Update needed for using grooming techniques at HLT (previously incompatible with confdb). A brief summary of this PR:
- Update FastjetJetProducer, and the classes affected by this change, with fillDescriptions method.
as discussed with @rappoccio @ferencek @ahinzmann @nhanvtran @jdolen @gkasieczka

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by alefisico** [comment]
The error in the runTheMatrix is related with the change itself. 
For instance, here: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18832/19968/runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step2_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log
FastjetJetProducer is NOT using the variables at all (they were defined but not using anywhere): doOutputJets or sumRecHits, but in the HLT_GRun_cff.py, those parameters are included in all the FasjetJetProducer module.
In my internal runTheMatrix test, when I commented those two variables from the modules, all the tests passed.

**Comment by perrotta** [comment]
If you move to a parameter set description and validation, you must expect that unknown parameters result into an error: that is one of the (good) reasons for using the fillDescriptions method, in fact.

Until the new parameter set is parsed in ConfDB and then propagated back in CMSSW, please add a customization in 
HLTrigger/Configuration/python/customizeHLTforCMSSW.py
so that this PR can be merged and successfully run in current CMSSW.

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 1]
```diff
@@ -165,6 +165,13 @@ def customiseFor18429(process):
 
      return process
 
+# Updating FatjetJetProducer with fillDescriptions
+def customiseFor18832(process):
+    for producer in producers_by_type(process, "FastjetJetProducer"):
+        if not hasattr(producer, "doOutputJets"): producer.doOutputJets = cms.bool(False)
+        if not hasattr(producer, "sumRecHits"): producer.sumRecHits = cms.bool(False)
```
Comment (perrotta) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: Did you try this customization? I think you must do the contrary, instead: if there is that parameter in the config, you must remove it. The syntax should be

     if hasattr(producer, "doOutputJets"): del producer.doOutputJets
     if hasattr(producer, "sumRecHits"): del producer.sumRecHits

But please check yourself if it works before re-committing

@Martin-Grunewald

**Comment by ferencek** [comment]
@alefisico, haven't you just done the exact opposite of what is needed. Instead of removing you added those two parameters to all instances of the FastJetProducer.

My suggestion for this PR was to introduce minimal changes to the code by defining a bit more complicated ParameterSetDescription with sets of optional parameters, etc. (something like https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_3/RecoBTag/SecondaryVertex/plugins/TemplatedSecondaryVertexProducer.cc#L1220-L1313). During the discussion we had I understood that this would not be possible for some reason (why not if it works fine for b tagging?) and that the HLT would be starting from a template configuration automatically generated from the ParameterSetDescription defined in the fillDescriptions method (how if there is only the default description that does not trigger the generation of a cfi file?). We now have a PR with a lot of changes which are not even displayed by GitHub and there are no unit tests that are testing all possible parameter combinations to ensure that no unintended changes were introduced.

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py (hunk 1 above)
Comment (alefisico) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: ups, yeah sorry. I just tried to fix it quickly. I will run the runTheMatrix now with this changes before I commit the next change.

**Comment by alefisico** [comment]
@ferencek for the first comment: yeah yeah sorry, I tried to do it faster and I didn't put much thought on it. I test the runTheMatrix before I commit the next change.
About the second: the changes can be displayed by github if you click in load diff. You will see that the main part of the code is untouched, the changes are only in the constructor and fillDescriptions.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ferencek** [comment]
@alefisico, sure, the diff can be loaded but large blocks of code have been replaced so by eye it is very hard to tell everything is ok. So we have to rely on automatic tests but they are not fully exhaustive.

**Comment by alefisico** [comment]
I just run runTheMatrix.py -l limited -i all and all the test passed:
https://cernbox.cern.ch/index.php/s/I0gm9C9fJvFL3CU

**Comment by perrotta** [comment]
please test

**Comment by alefisico** [comment]
Fixed RecoHi and also run the runTheMatrix test successfully: https://cernbox.cern.ch/index.php/s/7KhoarZk2NPbBLx

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoJets/JetProducers/plugins/VirtualJetProducer.cc [hunk 2]
```diff
@@ -228,8 +228,6 @@ VirtualJetProducer::VirtualJetProducer(const edm::ParameterSet& iConfig) {
 		fjRangeDef_ = RangeDefPtr( new fastjet::RangeDefinition(rhoEtaMax) );
 	} 
 
-	if( doFastJetNonUniform_ ) puCenters_ = iConfig.getParameter<vector<double> >("puCenters"); 	/// where is this used?
```
Comment (perrotta) [inline RecoJets/JetProducers/plugins/VirtualJetProducer.cc]: This is used, for example, at lines 566 and 578 of this file: you can't get rid of it as such!
(Probably the comment in the original code wanted to mean "Who's using the flag doFastJetNonUniform_?". But I'm just guessing...

Comment (alefisico) [inline RecoJets/JetProducers/plugins/VirtualJetProducer.cc]: yes, you are right. I'll fix it

Comment (alefisico) [inline RecoJets/JetProducers/plugins/VirtualJetProducer.cc]: ok, done. Here is the runTheMatrix file: https://cernbox.cern.ch/index.php/s/AZxTXWLOsyxWtaO

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
This PR has now merge conflicts. Furthermore, some additional parameters must be removed or
marked as deprecated so that their existence does not create this problem
```
----- Begin Fatal Exception 20-May-2017 10:42:21 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=FastjetJetProducer label='akPu3CaloJets'
Exception Message:
Illegal parameters found in configuration.  The parameters are named:
 'addNegative'
 'addNegativesFromCone'
 'dropZeros'
 'infinitesimalPt'
 'sumRecHits'
You could be trying to use parameter names that are not
allowed for this plugin or they could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

**Comment by alefisico** [comment]
Hi @Martin-Grunewald sorry, which test can I run locally in order to test the commit before I push it? I ran: "runTheMatrix.py -l limited -i all" and all the test passed: https://cernbox.cern.ch/index.php/s/AZxTXWLOsyxWtaO

**Comment by Martin-Grunewald** [comment]
To run the specific test:
```
  runTheMatrix.py --useInput=all --list=140.54
  ```
and
```
 addOnTests.py
```
after cmsenv inside a developer area.

**Comment by alefisico** [comment]
@Martin-Grunewald I am having an issue with one of the parameters. I ran the runTheMatrix as you mentioned, I am getting this error:
----- Begin Fatal Exception 22-May-2017 22:06:42 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=FastjetJetProducer label='akPu3CaloJets'
Exception Message:
MissingParameter: Parameter 'sumRecHits' not found.
----- End Fatal Exception -------------------------------------------------
but when I check the module ak4Pu2CaloJets using python -i, I see that no 'sumRecHits' is present:
>>> process.akPu3CaloJets
cms.EDProducer("FastjetJetProducer",
    Active_Area_Repeats = cms.int32(1),
    GhostArea = cms.double(0.01),
    Ghost_EtaMax = cms.double(6.5),
    Rho_EtaMax = cms.double(4.4),
    addNegative = cms.bool(True),
    addNegativesFromCone = cms.bool(False),
    doAreaDiskApprox = cms.bool(False),
    doAreaFastjet = cms.bool(True),
    doFastJetNonUniform = cms.bool(True),
    doPUOffsetCorr = cms.bool(True),
    doPVCorrection = cms.bool(False),
    doRhoFastjet = cms.bool(False),
    dropZeros = cms.bool(True),
    infinitesimalPt = cms.double(0.005),
    inputEMin = cms.double(0.0),
    inputEtMin = cms.double(0.3),
    jetAlgorithm = cms.string('AntiKt'),
    jetPtMin = cms.double(10.0),
    jetType = cms.string('CaloJet'),
    maxBadEcalCells = cms.uint32(9999999),
    maxBadHcalCells = cms.uint32(9999999),
    maxProblematicEcalCells = cms.uint32(9999999),
    maxProblematicHcalCells = cms.uint32(9999999),
    maxRecoveredEcalCells = cms.uint32(9999999),
    maxRecoveredHcalCells = cms.uint32(9999999),
    minSeed = cms.uint32(14327),
    nExclude = cms.uint32(2),
    nSigmaPU = cms.double(1.0),
    puCenters = cms.vdouble(-5, -4, -3, -2, -1, 
        0, 1, 2, 3, 4, 
        5),
    puPtMin = cms.double(6),
    puWidth = cms.double(0.8),
    rParam = cms.double(0.3),
    radiusPU = cms.double(0.5),
    src = cms.InputTag("towerMaker"),
    srcPVs = cms.InputTag("offlinePrimaryVertices"),
    subtractorName = cms.string('MultipleAlgoIterator'),
    useDeterministicSeed = cms.bool(True),
    voronoiRfact = cms.double(-0.9)
)

In fact, I am removing any sumRecHits from the process here: https://github.com/cms-sw/cmssw/pull/18832/commits/2bdbf73d43cf81d07b8a36e77d1efaee4697dc15#diff-5b29e53ee0c1d396908444c2711d39f6R172

any recommendation?

**Comment by Martin-Grunewald** [comment]
In your developer area, did you recompile everything with scram
after doing:
```
  git cms-checkdeps -A -a
```
to get all the dependent packages?
Did you check all compilation was successfull?
As the PR has merge conflicts, there MUST be a problem somewhere during compilation....

**Comment by alefisico** [comment]
Closing this PR, superseed by #18975

**Pull request closed without merging**